Add "distance spacing" support in editor for osu! ruleset

frenzibyte opened the pull request:
- [x] Depends on #16574 

Mostly resolves #16445 minus the osu!catch part.

https://user-images.githubusercontent.com/22781491/150663489-3017cd0f-cc9f-45a8-988f-9d4981fe45ad.mp4

Few notes:

- I'm mostly fine with using acronyms for the contracted display (`D. S.`), but could instead use abbreviations (`Dist. Spc.`).

- ~~osu!(stable) assigns the distance spacing hotkey to <kbd>Alt</kbd> and uses scroll to increase/decrease the value (<kbd>Alt</kbd>+<kbd>Wheel</kbd>), but that's what osu!(lazer) uses to increase/decrease the game volume. Therefore I added <kbd>D</kbd> to the combination to work with that (maybe I shoud look at defining a "distance spacing slider activation" keybinding instead?).~~ Fixed by using <kbd>Ctrl</kbd>+<kbd>Alt</kbd> for the time being.

- ~~(macOS-specific) apparently, holding shift while scrolling changes the direction from vertical to horizontal on an OS level (if I'm testing correctly), which makes holding shift while changing distance spacing via hotkey scorlling not actually work on macOS. I think we better want this fixed at osu!framework side via native cocoa API somehow instead. Queued it up on my list.~~ Removed functionality for the time beingm upon discussion.

- The expanding container (`SnappingToolboxContainer`)'s contracted and expanded were ballparked from what looks narrow/wide enough for the distance spacing slider. (although the settings content padding makes it a bit more pushed to the right, unfourtnately...)

- I'm not sure about the way `DistanceSnapGrid` is updated from distance spacing changes (by exposing the distance spacing bindable and making the grid listen for value changes), but I went with that in a similar fashion to how `BeatDivisor` is exposed in `IBeatSnapProvider` (although that one is exposed as a value). Alternatively I was thinking of exposing a `BeatSnapDistanceChanged` event that would be invoked when either `BeatDivisor` or `DistanceSpacing` has changed, but I believe a bindable is a better choice against an event.

frenzibyte committed:
Add "distance spacing" multiplier for osu! ruleset

While osu!catch also implements a distance snap grid, it doesn't rely on
`GetBeatSnapDistanceAt` (unlike osu!), therefore it can't have the
"distance spacing" multiplier yet.

frenzibyte committed:
Expose distance spacing in `IPositionSnapProvider` for updating distance grid

Alternate method is to expose a `SnapDistancesChanged` event in
`IPositionSnapProvider` instead, but I chose this way as an analogue to
`IBeatSnapProvider.BeatDivisor`, which might even make sense to be
exposed as `BindableBeatDivisor` instead of caching that separately.

frenzibyte committed:
Add test case for distance spacing hotkeys scrolling behaviour

frenzibyte committed:
Update failing snap test scenes

frenzibyte committed:
Remove `e.ShiftPressed` handling for now

Broken on macOS, will be handled differently later on as discussed.

frenzibyte committed:
Remove test coverage of "shift" hotkey behaviour

frenzibyte committed:
Merge branch 'master' into osu-distance-spacing

peppy commented:
Test failures look relevant.

frenzibyte committed:
Remove no longer required property

frenzibyte committed:
Allow using `SizeSlider` on different value types

frenzibyte committed:
Change "distance spacing" multipler type to `double`

Avoids losing precision on initial load, causing an unnecessary
hash change in `EditorChangeHandler`.

Resolves test failures in `TestSceneEditorChangeStates`
(https://github.com/ppy/osu/runs/5192493482?check_suite_focus=true).

frenzibyte committed:
Update outdated bindable casts in tests

peppy commented:
Are you able to explain how the recent commits fixed the test failures? Seems like the undo/redo test was failing originally, so I'd preferably like to know how having the wrong type on the slider caused such failures.

peppy commented on osu.Game/Overlays/Settings/Sections/SizeSlider.cs:
```diff
@@ -9,8 +10,9 @@ namespace osu.Game.Overlays.Settings.Sections
+        where T : struct, IEquatable<T>, IComparable<T>, IConvertible, IFormattable
     {
-        public override LocalisableString TooltipText => Current.Value.ToString(@"0.##x");
+        public override LocalisableString TooltipText => Current.Value.ToString(@"0.##x", null);
```
What is this `null`?

frenzibyte replied:
A value to use the current culture info, according the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.iformattable.tostring?view=net-6.0#system-iformattable-tostring(system-string-system-iformatprovider)):

![CleanShot 2022-02-16 at 10 24 53](https://user-images.githubusercontent.com/22781491/154216064-8f86fdc7-b63f-41cc-868f-d6085e89ce47.png)

Can replace with `NumberFormatInfo.CurrentInfo` if that reads better.

peppy replied:
Please do. null looks very very weird.

peppy commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -56,6 +58,26 @@ public abstract class HitObjectComposer<TObject> : HitObjectComposer, IPlacement
+        /// <remarks>
+        /// Setting this to <see langword="true"/> displays a "distance spacing" slider and allows this composer to configure the value of <see cref="BeatmapInfo.DistanceSpacing"/>.
+        /// </remarks>
+        protected abstract bool SupportsDistanceSpacing { get; }
```
Was there a reason for doing it this way, rather than having it a per-ruleset implementation?

frenzibyte replied:
I can see it being shared across osu! and osu!catch, but I think I could move out the whole distance spacing handler to its own component and use it in each ruleset implementation.

peppy replied:
Not 100% sure yet. Let me dwell on this a bit longer.

bdach replied:
I think I would personally slightly prefer not having this here and this being attached in rulesets in a self-contained way if possible. I'm worried that it will become a precedent for all sorts of abstract flags and switches on `HitObjectComposer` and I'm not at all sure that it's a path we want to be taking.

frenzibyte commented:
That undo/redo test was failing on an assert step ("no changes") which ensures `Editor.HasUnsavedChanges = false`. 

I have mentioned it in the commit description of https://github.com/ppy/osu/pull/16576/commits/0992bec2c849d9e5a7be97d1d9396027a1f8f4e1, but basically, this is how the distance spacing bindable used to work:
https://github.com/ppy/osu/blob/19ee05c232eb86710001bdad3bef2e135f7bc97a/osu.Game/Rulesets/Edit/HitObjectComposer.cs#L164

https://github.com/ppy/osu/blob/19ee05c232eb86710001bdad3bef2e135f7bc97a/osu.Game/Rulesets/Edit/HitObjectComposer.cs#L184-L189

That potentially causes precision loss and messes with the current hash state of `EditorChangeHandler`, making `Editor.HasUnsavedChanges` become true and cause that test to fail on the "no changes" assert step.

peppy commented:
Ignoring questionable decisions, seems to work as expected.

frenzibyte committed:
Use explicit culture info rather than `null`

bdach commented on osu.Game/Rulesets/Edit/IPositionSnapProvider.cs:
```diff
@@ -1,13 +1,21 @@
 {
     public interface IPositionSnapProvider
     {
+        /// <summary>
+        /// The spacing multiplier applied to beat snap distances.
+        /// </summary>
+        /// <seealso cref="BeatmapInfo.DistanceSpacing"/>
+        IBindable<double> DistanceSpacingMultiplier { get; }
```
Not entirely certain this belongs to the interface. Position snap is more general than distance snap, so I imagine it may not make sense to implement this in every single ruleset (in fact mania and taiko can't ever do anything useful with this, while they generally can perform position snap, just a _different_ sort of snap).

Thoughts on splitting off `IDistanceSnapProvider : IPositionSnapProvider`, implementing it only where it makes sense (osu! and catch), and resolving it via the more-specific type where it makes sense? Haven't at all checked if it's viable. just spitballing for now.

bdach replied:
Also, note that if the above happens, you can potentially remove `SupportsDistanceSpacing` as a property entirely and check if a non-null `IDistanceSnapProvider` has been cached instead in order to add the distance spacing controls.

frenzibyte replied:
Seems good – there are other methods in `IPositionSnapProvider` which make sense to be moved to `IDistanceSnapProvider`, but I would rather keep it there for now as to not complicate this PR. 

Have added a TODO for it at least.

bdach commented on osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs:
```diff
@@ -32,6 +32,8 @@ public class CatchHitObjectComposer : HitObjectComposer<CatchHitObject>
 
         private readonly Bindable<TernaryState> distanceSnapToggle = new Bindable<TernaryState>();
 
+        protected override bool SupportsDistanceSpacing => false;
```
I've already mentioned that I ideally wouldn't want this switch to exist, but if it has to, then maybe it should be `false` by default rather than fully abstract and having to be overridden in each consumer? Not sure. It's a ruleset API breaking change as is for no real good reason.

frenzibyte committed:
Merge branch 'master' into osu-distance-spacing

frenzibyte committed:
Move `DistanceSpacingMultiplier` bindable to its own `IDistanceSnapProvider` interface

frenzibyte committed:
Encapsulate distance spacing control handling to a "distance toolbox composite"

Encapsulated in a way which can allow further extensibility for the
right-side area of toolboxes.

frenzibyte commented:
Took a good breather from this to refresh my mind. Have moved the distance spacing logic from `HitObjectComposer` to the osu! ruleset in a `OsuToolboxComposite` class.

The point of such class is to allow it take full size of the compose screen to allow for handling distance spacing scrolling shortcut (<kbd>Alt</kbd>+<kbd>D</kbd> + scroll), and potentially other more. I haven't gone with adding any abstraction layers to that composite class to keep things simplified, but I can see it moved to the base class and overridable by rulesets for specific controls. 

I have also added a TODO there to share implementation with catch once it supports distance spacing as well, along with another TODO on `CatchHitObjectComposer` to inherit from `IDistanceSnapProvider` as well.

frenzibyte committed:
Give snapping toolbox enough hover delay for interaction with expand button

With 0 delay, if the `SettingsToolboxGroup` has been contracted via
button, there's no chance to be able to expand it via button again.

This is temporary at best just to keep things working somewhat, not sure
if the expansion logic will still be required in editor redesign, so
deciding to roll with it for now.

peppy commented:
Not sure what's going on but the snapping doesn't seem to always work:

https://user-images.githubusercontent.com/191335/165273106-7507bd5a-6362-4e8a-9e89-16fc9f47a829.mp4

peppy commented on osu.Game.Rulesets.Osu/Edit/OsuToolboxComposite.cs:
```diff
@@ -0,0 +1,121 @@
+
+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            if (e.AltPressed && e.Key == Key.D && !e.Repeat)
```
I'm not sure where this shortcut comes from, but it's not correct. You can't have Alt+Alpha+Mouse input, it breaks unspoken UX rules.

frenzibyte replied:
For sure, this is open for suggestions. Stable uses Alt+Scroll for modifying the distance spacing, but as you know, that's a global-wide shortcut for modifying the game value.

peppy replied:
Yeah, I'm not yet sure on the best way to handle this. We already overload the mouse wheel too much. For the sake of keeping things moving forward, you could make it ctrl+alt for now?

frenzibyte committed:
Change distance spacing keybind to Ctrl+Alt

bdach commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -383,7 +385,7 @@ public override SnapResult SnapScreenSpacePositionToValidTime(Vector2 screenSpac
         public override float GetBeatSnapDistanceAt(HitObject referenceObject)
         {
-            return (float)(100 * EditorBeatmap.Difficulty.SliderMultiplier * referenceObject.DifficultyControlPoint.SliderVelocity / BeatSnapProvider.BeatDivisor);
+            return (float)(100 * referenceObject.DifficultyControlPoint.SliderVelocity * EditorBeatmap.Difficulty.SliderMultiplier / BeatSnapProvider.BeatDivisor);
```
What is this change even? It's just a change in the multiplication order that's wedged in here for no discernible reason...?

frenzibyte replied:
Probably leftover from the beginnings of the PR, can revert I guess.

bdach commented on osu.Game.Tests/Editing/TestSceneHitObjectComposerDistanceSnapping.cs:
```diff
@@ -61,6 +61,7 @@ public TestSceneHitObjectComposerDistanceSnapping()
             BeatDivisor.Value = 1;
 
             composer.EditorBeatmap.Difficulty.SliderMultiplier = 1;
+            composer.EditorBeatmap.BeatmapInfo.DistanceSpacing = 1;
```
Incidentally, why isn't `BeatmapInfo.DistanceSpacing` initialised by default to 1 in the model class itself? Seems like a better idea than patching readers all over.

frenzibyte replied:
I'm curious if its the default value and not because of the test beatmap. Will see.

frenzibyte replied:
There are three cases of this currently:
 - [`TestSceneObjectObjectSnap`](https://github.com/ppy/osu/pull/16576/files#diff-ba9e43f7e802590bea73bf4272d1f8318c3e94b868927ec612d0cb3485280c40) and [`TestSceneSliderSnapping`](https://github.com/ppy/osu/pull/16576/files#diff-fca3d8bdcf3b7b013aafcb562ebb2743e1d9a25029f5334401e38a563de58dd5), which awaits decision at https://github.com/ppy/osu/pull/16576#discussion_r860450100. 
 - [`TestSceneHitObjectComposerDistanceSnapping`](https://github.com/ppy/osu/pull/16576/files#diff-2d74227b54f85dd4814534d6da5bfc90cc1a69f57ad9cc5ba1b1fe14b2db948a), which has been intentionally left inline with the pre-existing `BeatDivisor = 1` and `SliderMultiplier = 1`, unless we want to mutate the values in `TestBeatmap` itself which I'm not 100% sure about (would basically break how the beatmap was mapped).

peppy replied:
I think `1` should be the default value in all cases. `0` doesn't make any sense.

peppy replied:
Misunderstood this. It seems that the actual test beatmap ("My Love") provides a non-default value. In which case I think this is fine (maybe include an inline comment noting this).

bdach commented on osu.Game.Rulesets.Osu/Edit/OsuToolboxComposite.cs:
```diff
@@ -0,0 +1,120 @@
+    /// <summary>
+    /// A toolbox composite for osu!-specific controls.
+    /// </summary>
+    // todo: once catch supports distance spacing, the control here should move out to a base "DistancingRulesetToolboxComposite" class or something better.
```
Strongly disagree with this todo. I'd much rather have this composite duplicated in catch than extract enterprise-grade inflexible base classes.

frenzibyte replied:
I don't know about that, I could look at splitting the slider itself to avoid a bit of duplication, but the distance spacing scroll logic can't be shared without sharing this class itself. And basically doing the same for custom rulesets as well.

frenzibyte replied:
See https://github.com/ppy/osu/pull/16576#discussion_r860370075, `OsuToolboxComposite` has just been removed in favour of `DistancedHitObjectComposer` which contains the distance spacing logic in there.

bdach commented on osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs:
```diff
@@ -128,6 +135,9 @@ protected override void Update()
         }
 
+        public override float GetBeatSnapDistanceAt(HitObject referenceObject)
+            => (float)(base.GetBeatSnapDistanceAt(referenceObject) * DistanceSpacingMultiplier.Value);
```
Cast is really unfortunate but I guess `DistanceSpacingMultiplier` was already a double before this branch, so I guess this'll have to live for the time being...

frenzibyte replied:
Yeah, distances consistently follow the float data type, so distance spacing was probably supposed to be float, but unsure if it's simple changing that in the model class.

peppy replied:
I believe realm stores all number types internally using the same storage type. Should be a simple change, you're welcome to try it (may not even require a migration).

bdach commented on osu.Game/Screens/Edit/Compose/Components/DistanceSnapGrid.cs:
```diff
@@ -82,6 +85,9 @@ protected override void LoadComplete()
             beatDivisor.BindValueChanged(_ => updateSpacing(), true);
+
+            distanceSpacingMultiplier = SnapProvider.DistanceSpacingMultiplier.GetBoundCopy();
+            distanceSpacingMultiplier.BindValueChanged(_ => updateSpacing(), true);
```
`beatDivisor` is already calling `updateSpacing()` once above via the `BindValueChanged(..., true)` call. Probably need just one of those spacing updates.

frenzibyte commented on osu.Game/Rulesets/Edit/IPositionSnapProvider.cs:
```diff
@@ -30,6 +30,8 @@ public interface IPositionSnapProvider
         /// </summary>
         /// <param name="referenceObject">An object to be used as a reference point for this operation.</param>
         /// <returns>The distance between two points residing in the timing point that are one beat length apart.</returns>
+        // todo: this, and other distance-related methods below it, should move to IDistanceSnapProvider.
```
I can look into it as a follow-up to this PR, and if everyone here agrees on this split.

frenzibyte replied:
Hmm, actually I'm not entirely sure how this can be properly work. The only idea I have for this is creating a separate `DistancedHitObjectComposer` which implements this interface, but does that seem like a good direction?

frenzibyte replied:
Have pushed changes since that resolves both this conversation and the one about the existence of `OsuToolboxComposite` as well.

frenzibyte committed:
Move distancing methods from `IPositionSnapProvider` to `IDistanceSnapProvider`

frenzibyte committed:
Introduce `DistancedHitObjectComposer` and supersede `OsuToolboxComposite`

frenzibyte committed:
Let `CatchHitObjectComposer` inherit from `DistancedHitObjectComposer` and hide control

frenzibyte commented on osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderSnapping.cs:
```diff
@@ -61,9 +63,14 @@ public override void SetUpSteps()
+                var beatDivisor = Editor.Dependencies.Get<BindableBeatDivisor>();
                 beatDivisor.Value = 1;
             });
+            AddStep("set distance spacing to 1", () =>
+            {
+                var distanceSpacing = (BindableDouble)Editor.ChildrenOfType<OsuHitObjectComposer>().Single().DistanceSpacingMultiplier;
+                distanceSpacing.Value = 1;
+            });
```
While looking at each case of this, I've realized this is actually hiding a change in behaviour between osu!(stable) and osu!(lazer):

In osu!(stable), the distance spacing multipler does not affect the snapping of the slider:

https://user-images.githubusercontent.com/22781491/165670957-d6eed6e1-ebff-4cf2-afb8-16d9ce1d825b.mp4

Meanwhile on osu!(lazer) with this change, it does affect the slider snapping:

https://user-images.githubusercontent.com/22781491/165670990-cfb81127-768b-49cd-b02a-b61daf08c4f2.mp4

I'm not entirely sure which makes more sense here, I would think osu!(lazer) is better since it's making snap distances consistent across hitcircles and sliders? maybe, or maybe that doesn't make sense (cc @peppy).

peppy replied:
Distance snap should only be applied to the head of a slider (ie. before the slider path placement begins). In your video, it looks like lazer is incorrectly attempting to distance snap the tail to the head or something? Shouldn't be a thing.

Slider placement is only affected by the beat snap divisor and slider velocity. Distance snap is specifically for the distance between adjacent hitobjects. To explain another way, with distance snap set to 1.0x, objects should be spaced apart at the same rate as a slider's ball moves down the path.

frenzibyte committed:
Keep distance spacing reset to one on test scenes with slider snapping for now

frenzibyte committed:
Add test case for `DistanceSpacingMultiplier` in distance snapping test scene

frenzibyte committed:
Set default value of `DistanceSpacing` to 1

frenzibyte committed:
Minor cleanup

Just from reading diff.

peppy commented on osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs:
```diff
@@ -42,6 +42,10 @@ public CatchHitObjectComposer(CatchRuleset ruleset)
         [BackgroundDependencyLoader]
         private void load()
         {
+            // todo: enable distance spacing once catch supports it.
```
May want to reword this comment to mention that the part which isn't supported by the osu!catch editor isn't distance spacing itself, but the ability to adjust the applied amount.

frenzibyte committed:
Convert data type of `DistanceSpacing` to float

frenzibyte committed:
Move distance spacing application to `DistanceSnapGrid`

frenzibyte committed:
Clarify TODO comment

bdach commented:
Tests are looking rather red here.

frenzibyte commented:
I fixed some of the test failures locally, but actually got stumbled with yet another issue due to the float-double switch, which is basically the following in short:

https://user-images.githubusercontent.com/22781491/165861600-3e9912f4-3d75-4d4d-84dd-d094073ade21.mp4

Causing the beatmap file to change from `DistanceSpacing: 1.2` to `DistanceSpacing: 1.1999999`.

The specific reason why this is happening is because of the distance spacing bindable having a precision of `0.01f`, which causes the original value (`1.2000005f`) to change to (`1.1999993f`) during the [precision logic](https://github.com/ppy/osu-framework/blob/313943f2d0ce2791f8c85ba9ca1417e8ee193708/osu.Framework/Bindables/BindableNumber.cs#L76-L77).

This can be fixed by adding an `AlmostEquals` guard before changing the value in the beatmap model, but I'm tempted to revert and avoid another issue with this...

bdach commented:
I'd say it's probably best to revert for now and maybe try again in another pull, with due care. I have worries that that type change is going to potentially affect beatmap conversions now, so that would need to be checked before attempting again.

frenzibyte committed:
Revert "Convert data type of `DistanceSpacing` to float"

This reverts commit 7aaa88cac2be239dcc778a1082585f8e8d5a7e55.

frenzibyte committed:
Add back `float` cast to distance spacing computation

bdach commented:
I've had a read through the newest changes and I feel like this pull has cleaned up rather nicely. About the only thing I'm uncertain about is whether `DistancedHitObjectComposer` is going to hold up in the long run, and even then it's not *too* bad right now, I don't think.

This is too large a change to merge on my own, so going to wait for a second review here, but as far as I'm concerned this looks ready.

peppy commented:
One thing that's missing is exposing the shortcut to adjust it. I wonder if we should put these in the tooltips on the sliders until we have a better place for them?

Optimally, these settings would either show in the global OSD, or in an editor local one if that is too hard to implement. Haven't looked into what it would take to make this happen (likely best left as a separate PR if the direction is agreed upon).

frenzibyte commented:
Will look into that in a follow-up PR.
